src/server_manager/electron_app/digitalocean_oauth.ts

```diff
@@ -0,0 +1,216 @@
+  let portIdx = 0;
+  return new Promise((resolve, reject) => {
+    server.once('listening', () => {
+      console.log(`Listening on port ${portList[portIdx]}`);
+      resolve(portIdx);
+    });
+    server.on('error', (error: ServerError) => {
+      if (error.code === 'EADDRINUSE') {
+        console.log(`Port ${portList[portIdx]} already in use`);
+        portIdx += 1;
+        if (portIdx < portList.length) {
+          const port = portList[portIdx];
+          console.log(`Trying port ${port}`);
+          server.listen({host: 'localhost', port, exclusive: true});
+          return;
+        }
+      }
+      server.close();
+      reject(error);
+    });
+    server.listen({host: 'localhost', port: portList[portIdx], exclusive: true});
+  });
+}
+
+// See https://developers.digitalocean.com/documentation/v2/#get-user-information
+interface Account {
+  droplet_limit: number;
+  floating_ip_limit: number;
+  email: string;
+  uuid: string;
+  email_verified: boolean;
+  status: string;
+  status_message: string;
+}
+
+function getAccount(accessToken: string): Promise<Account> {
+  return new Promise((resolve, reject) => {
+    request(
+        {
+          url: 'https://api.digitalocean.com/v2/account',
+          headers: {
+            'User-Agent': 'Outline Manager',
+            'Content-Type': 'application/json',
+            'Authorization': `Bearer ${accessToken}`
+          }
+        },
+        (error, response, body) => {
+          if (error) {
+            return reject(error);
+          }
+          const bodyJson: {account: Account} = JSON.parse(body);
+          return resolve(bodyJson.account as Account);
+        });
+  });
+}
+
+export interface OauthSession {
+  // The result of the OAuth session, with the authentication token.
+  result: Promise<string>;
+  // Checks if the OauthSession has been cancelled.
+  isCancelled(): boolean;
+  // Cancels the OAuth session. The result promise is automatically rejected
+  // and isCancelled() will return true.
+  cancel(): void;
+}
+
+// Runs the DigitalOcean oauth flow and returns the access token.
+// See https://developers.digitalocean.com/documentation/oauth/ for the API.
+export function runOauth(): OauthSession {
+  const secret = randomValueHex(16);
+
+  const app = express();
+  const server = http.createServer(app);
+  server.on('close', () => console.log('Oauth server closed'));
+
+  let isCancelled = false;
+  app.use((req, res, next) => {
+    if (isCancelled) {
+      res.status(503).send('Authentication cancelled');
+    } else {
+      next();
+    }
+  });
+
+  app.get('/', (request, response) => {
+    response.send(`<html>
+          <head><title>Authenticating...</title></head>
+          <body>
+              <form id="form" method="POST">
+                  <input id="params" type="hidden" name="params"></input>
+              </form>
+              <script>
+                  let params = new URLSearchParams(location.hash.substr(1));
```

Comment:
It seems like this javascript's purpose is to extract information from the URL fragment and send it to the server.  Can we simplify this by putting the information in actual query parameters instead?

Reply:
Unfortunately the DO OAuth flow puts the access token in the hash. That's why we need the javascript. It's a workaround.